### DISCUSSION TITLE
Batching Dicts

### DISCUSSION THREAD

**Discussion by vkovac2** [pyg-team/pytorch_geometric#7622] [Q&A] [answered]
Hello everyone.

 I am trying to make a transformation dataset class that will be used in “pre_transform” of QM9 dataset, and I have came up with one issue  when dealing with dictionaries. When having  nodes of two different node types we can use the “__inc__” method as described in “Advanced Mini-Batching” post. However, is there a way to deal with dictionaries in a similar manner? So if I have a dictionary that matches an id with an edge {0:[0,1], 1:[2,3]...}, is it possible to make it work for batches later somehow? The problem that arises is that the values in the dictionary won’t get incremented when constructing a batch, thus leading to confusions. Is there a way to overcome this issue when constructing custom dataset transformation class?

Thank you very much for your time in advance

**Comment by rusty1s** [comment]
Yes, that is possible. You can see that `__inc__` gets called here for every key in the dictionary:
```python
import torch

from torch_geometric.data import Batch, Data


class MyData(Data):
    def __inc__(self, key, value, *args, **kwargs):
        print(key, value)
        return 0


data = MyData(x_dict={'1': torch.tensor([0]), '2': torch.tensor([1])})

batch = Batch.from_data_list([data, data])
print(batch)

**Reply by vkovac2** [reply]
Hello. Thanks a lot for your quick reply. I am not sure whether I explained my problem correctly. I attached an image so I am more clear in my explanation. Here the batch_size is 128. The "batch.edges" is basically a a list of values coming from a dictionary where id->edge (node pair) is stored ("edges":list(cc.edge_lookup.values()), where edge_lookup is the id->edge), and "adj_00" is basically the "edge_index" of QM9 dataset. So, as you can see from the picture the adjacency gets incremented correctly when batching using your techniques in the post of "Advanced Mini-Batching", so it starts with node id 0 and goes all the way to 1292. However, the problem is the increment in the values of batch.edges, which was originally the dictionary. So ideally here I would like the last entries of batch.edges[127] to be equal to [1292,1285], but instead I am getting again [0,1]. Is there a way to do that?

Thanks a lot

<img width="864" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/49810042/cae83c88-eb91-4171-a64f-0859ac9187a9">

**Reply by rusty1s** [reply]
Yes, I think you can do this by adjusting the `__inc__` method. For example,
```python
import torch

from torch_geometric.data import Batch, Data


class MyData(Data):
    def __cat_dim__(self, key, value, *args, **kwargs):
        if key == '0':
            return 1
        if key == '1':
            return 1
        return super().__cat_dim__(key, value, *args, **kwargs)

    def __inc__(self, key, value, *args, **kwargs):
        if key == '0':
            return 10
        if key == '1':
            return 20
        return super().__inc__(key, value, *args, **kwargs)


data = MyData(edges={
    '0': torch.tensor([[0], [0]]),
    '1': torch.tensor([[1], [1]]),
})

batch = Batch.from_data_list([data, data])
print(batch)
print(batch.edges)
```
```
MyDataBatch(
  edges={
    0=[2, 2],
    1=[2, 2],
  }
)
{'0': tensor([[ 0, 10], [ 0, 10]]),
 '1': tensor([[ 1, 21], [ 1, 21]])}

```

**Reply by vkovac2** [reply]
Thanks again for your prompt reply. I think that the issue I am describing comes from the cases when there are different dimensions in the dictionary. I am attaching a code snippet so it is more clear.

As you can see the adjacency matrix values are getting incremented as they should for the second case. The problem that I have is how can I increment in a similar manner the "list_relations" or "relations" dict. The problem is probably the fact that I cannot make it a tensor as the lists of dictionary "relations" are of different dimension. Is there any way to increment the values in those lists? A potential solution would probably be to create dummy values in order to make it as a tensor and then increment that. Is there any smarter way to do it?

```py
import torch

from torch_geometric.data import Batch, Data
from typing import Any


class MyData(Data):
    def __inc__(self, key, value, *args, **kwargs):
        if 'adj' in key: 
            return torch.tensor([[getattr(self, 'x').size(0)], [getattr(self, 'x').size(0)]])
        else:
            return super().__inc__(key, value, *args, **kwargs)
    def __cat_dim__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'adj' in key:
            return 1
        else:
            return 0
            

adj = torch.tensor([[0,  1,  1,  2,  2,  2,  3,  3,  4,  4,  5,  5,  5,  6,  6,  7,  7,  8,
          8,  8,  9, 10, 10, 10, 11, 11, 12, 12, 12, 13, 13, 14, 14, 15, 15, 15,
         16, 16, 16, 16, 17, 18, 19, 19, 19, 20, 20, 21, 21, 21, 22, 23, 23, 24,
         24, 25, 25, 26, 26, 27, 27, 27, 28, 28], 
         [ 1,  0,  2,  1,  3, 28,  2,  4,  3,  5,  4,  6, 27,  5,  7,  6,  8,  7,
          9, 10,  8,  8, 11, 27, 10, 12, 11, 13, 26, 12, 14, 13, 15, 14, 16, 25,
         15, 17, 18, 19, 16, 16, 16, 20, 24, 19, 21, 20, 22, 23, 21, 21, 24, 19,
         23, 15, 26, 12, 25,  5, 10, 28,  2, 27]])

relations = {0:[2, 28, 27, 10, 8, 7, 6, 5, 4, 3], 1:[2, 28, 27, 5, 4, 3], 2:[2, 3, 4, 5, 27, 28]}
list_relations = list(relations.values())


data = MyData(x = torch.randn([29,5]), adj = adj, list_relations = list_relations)

batch = Batch.from_data_list([data, data])
print(batch)

```

Thanks a lot again for your help

**Reply by rusty1s** [reply]
I think you can still increment them if values in `relations` are given as a PyTorch Tensor. Nonetheless, I think what you are actually looking for is [`torch.nested`](https://pytorch.org/docs/stable/nested.html).

**Reply by vkovac2** [reply]
That is amazing. That is exactly what I was looking for. Thanks a lot. After experimenting a bit however, I am getting the following error: 

Could not run 'aten::cat' with arguments from the 'NestedTensorCPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::cat' is only available for these backends: [CPU, CUDA, HIP, MPS, IPU, XPU, HPU, VE, Meta, MTIA, PrivateUse1, PrivateUse2, PrivateUse3, FPGA, ORT, Vulkan, Metal, QuantizedCPU, QuantizedCUDA, QuantizedHIP, QuantizedMPS, QuantizedIPU, QuantizedXPU, QuantizedHPU, QuantizedVE, QuantizedMeta, QuantizedMTIA, QuantizedPrivateUse1, QuantizedPrivateUse2, QuantizedPrivateUse3, CustomRNGKeyId, MkldnnCPU, SparseCPU, SparseCsrCPU, SparseCsrCUDA, BackendSelect, Python, FuncTorchDynamicLayerBackMode, Functionalize, Named, Conjugate, Negative, ZeroTensor, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradHIP, AutogradXLA, AutogradMPS, AutogradIPU, AutogradXPU, AutogradHPU, AutogradVE, AutogradLazy, AutogradMeta, AutogradMTIA, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, AutogradNestedTensor, Tracer, AutocastCPU, AutocastCUDA, FuncTorchBatched, FuncTorchVmapMode, Batched, VmapMode, FuncTorchGradWrapper, PythonTLSSnapshot, FuncTorchDynamicLayerFrontMode, PythonDispatcher].

Or when I put the increment that I want
``` py

    def __inc__(self, key, value, *args, **kwargs):
        if 'adj' in key: 
            return torch.tensor([[getattr(self, 'x').size(0)], [getattr(self, 'x').size(0)]])
        elif "list" in key:
            return torch.tensor([getattr(self, 'x').size(0)])
        else:
            return super().__inc__(key, value, *args, **kwargs)
```


 I get the following error:


"
Exception has occurred: RuntimeError
Expected both self and other to be nested, but got a nested self and non-nested other.
  File "/Users/try.py", line 38, in <module>
    batch = Batch.from_data_list([data, data])
"


I found out that PyG has its own function for that called "torch_geometric.utils.to_nested_tensor()", but it doesn't function for my code as well with the following error although the input is list of tensors as expected:

"Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/opt/anaconda3/envs/py3.9/lib/python3.9/site-packages/torch_geometric/utils/nested.py", line 9, in to_nested_tensor
    def to_nested_tensor(
  File "/Users/opt/anaconda3/envs/py3.9/lib/python3.9/site-packages/torch/nested/__init__.py", line 18, in as_nested_tensor
    def as_nested_tensor(
TypeError: nested_tensor(): Expected first argument to be a list of tensors "

Do you by any chance know how can I solve the above issues?

Thanks a lot again.

**Reply by rusty1s** [reply]
How do you call `to_nested_tensor` from PyG? I would need to investigate a bit how we can enable nested tensors in `Batch.from_data_list`. Will take a look.

**Reply by vkovac2** [reply]
Actually I do not use "Batch.from_data_list" in my actual code and it still doesn't work. This was just to explain my problem better. In the actual code I just use it similarly with a dataloader, but it never reaches that point as it throws an error earlier. 

A potential solution is the following: 
```py
list_relations = torch.nested.to_padded_tensor(torch.nested.nested_tensor(list(relations.values())), padding = -9999)
```

, but then I would have to remove the negative values from my batch in order to be able to index the respective positions.


The way I use the PyG "to_nested_tensor" is the following:
```py
import torch

from torch_geometric.data import Batch, Data
from typing import Any
from torch_geometric.utils import to_nested_tensor

class MyData(Data):
    def __inc__(self, key, value, *args, **kwargs):
        if 'adj' in key: 
            return torch.tensor([[getattr(self, 'x').size(0)], [getattr(self, 'x').size(0)]])
        elif "list" in key:
            return torch.tensor([getattr(self, 'x').size(0)])
        else:
            return super().__inc__(key, value, *args, **kwargs)
    def __cat_dim__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'adj' in key:
            return 1
        else:
            return 0
            


adj = torch.tensor([[0,  1,  1,  2,  2,  2,  3,  3,  4,  4,  5,  5,  5,  6,  6,  7,  7,  8,
          8,  8,  9, 10, 10, 10, 11, 11, 12, 12, 12, 13, 13, 14, 14, 15, 15, 15,
         16, 16, 16, 16, 17, 18, 19, 19, 19, 20, 20, 21, 21, 21, 22, 23, 23, 24,
         24, 25, 25, 26, 26, 27, 27, 27, 28, 28], 
         [ 1,  0,  2,  1,  3, 28,  2,  4,  3,  5,  4,  6, 27,  5,  7,  6,  8,  7,
          9, 10,  8,  8, 11, 27, 10, 12, 11, 13, 26, 12, 14, 13, 15, 14, 16, 25,
         15, 17, 18, 19, 16, 16, 16, 20, 24, 19, 21, 20, 22, 23, 21, 21, 24, 19,
         23, 15, 26, 12, 25,  5, 10, 28,  2, 27]])

relations = {0:torch.tensor([2, 28, 27, 10, 8, 7, 6, 5, 4, 3]), 1:torch.tensor([2, 28, 27, 5, 4, 3]), 2:torch.tensor([2, 3, 4, 5, 27, 28])}
list_relations = to_nested_tensor(list(relations.values()))

data = MyData(x = torch.randn([29,5]), adj = adj, list_relations = list_relations)

batch = Batch.from_data_list([data, data])
print(batch)

```


Thanks

**Reply by rusty1s** [reply, marked as answer]
Added this via https://github.com/pyg-team/pytorch_geometric/pull/7643

**Reply by vkovac2** [reply]
Amazing. Thank you very much. It is not possible still to increment the values in the nested tensor though right? Is there any way to overcome that issue?

**Reply by rusty1s** [reply]
That should work just fine by incrementing via `__inc__`.

**Reply by rusty1s** [reply]
Confirmed here: https://github.com/pyg-team/pytorch_geometric/pull/7647

**Reply by vkovac2** [reply]
Amazing. Thank you very much. Increment was not working last night bcs I had a tensor for its value, but I saw from your test code that it should be an integer when we have nested tensors. Thanks!

**Reply by vkovac2** [reply]
Hello again @rusty1s. I tried implementing the new feature that you created on ZINC dataset and it didn't work. It seems that the PyTorch "nested" does not support the "size" or the "cat" that advanced mini-batching of PyG expects. I am attaching a sample code that I created so you can understand better the problem. 


```py

import torch
from torch_geometric.data import Data
from torch_geometric.transforms import BaseTransform
from typing import Any
from torch_geometric.transforms import Compose
from torch_geometric.datasets import ZINC, QM9
from torch_geometric.loader import DataLoader

class customData(Data):

    @classmethod
    def take_feat(cls, data: Data):

        relations = {0:[2, 7, 6, 5, 4, 3], 1:[2, 5, 4, 3], 2:[2, 3, 4, 5, 1]}
        data["relations"] = torch.nested.nested_tensor(list(relations.values()))
        return data
    def __cat_dim__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'edge_index' in key:
            return 1
        else:
            return 0
        
    def __inc__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'edge_index' in key:
            return torch.tensor([[getattr(self, 'x').size(0)], [getattr(self, 'x').size(0)]])
        if 'relation' in key:
            return getattr(self, 'x').size(0)
        else:
            return super().__inc__(key, value, *args, **kwargs)

class customTransformations(BaseTransform):
    def __call__(self, data: Data):
        
        new_data = customData.take_feat(data)
        return new_data


data_train = ZINC("datasets/custom_zinc3", subset=True, split='train',
                    pre_transform=Compose([customTransformations()]), transform=Compose([]))  

train_loader = DataLoader(data_train, batch_size=128)
```


Error that I got: "Exception has occurred: RuntimeError
Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
  File "/Users/try_zinc.py", line 44, in <module>
    data_train = ZINC("datasets/custom_zinc3", subset=True, split='train',"
    
    
   
Any suggestions to overcome this issue?  Thanks a lot again

**Reply by rusty1s** [reply]
Will take a look soon.

**Reply by rusty1s** [reply]
This works just fine for me
```python
from typing import Any

import torch

from torch_geometric.data import Data
from torch_geometric.datasets import QM9, ZINC
from torch_geometric.loader import DataLoader
from torch_geometric.transforms import BaseTransform, Compose


class customTransformations(BaseTransform):
    def __call__(self, data: Data):
        relations = {
            0: [2, 7, 6, 5, 4, 3],
            1: [2, 5, 4, 3],
            2: [2, 3, 4, 5, 1]
        }
        data.relations = torch.nested.nested_tensor(list(relations.values()))
        return data


data_train = ZINC("/tmp/custom_zinc3", subset=True, split='train',
                  transform=Compose([customTransformations()]))

train_loader = DataLoader(data_train, batch_size=128)
batch = next(iter(train_loader))
print(batch)
```
Are you on latest PyG master?

**Reply by vkovac2** [reply]
Hey @rusty1s .The code snipped you posted work fine for me as well. However, when you add the increment method as on the code snippet I posted earlier, it does not work. Seems that it is something that hasn't been implemented yet from Pytorch.

Thanks

**Reply by rusty1s** [reply]
I cannot reproduce this:
```python
from typing import Any

import torch

from torch_geometric.data import Data
from torch_geometric.datasets import QM9, ZINC
from torch_geometric.loader import DataLoader
from torch_geometric.transforms import BaseTransform, Compose


class customData(Data):
    @classmethod
    def take_feat(cls, data: Data):

        relations = {
            0: [2, 7, 6, 5, 4, 3],
            1: [2, 5, 4, 3],
            2: [2, 3, 4, 5, 1]
        }
        data["relations"] = torch.nested.nested_tensor(list(
            relations.values()))
        return data

    def __cat_dim__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'edge_index' in key:
            return 1
        else:
            return 0

    def __inc__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'edge_index' in key:
            return torch.tensor([[getattr(self, 'x').size(0)],
                                 [getattr(self, 'x').size(0)]])
        if 'relation' in key:
            return getattr(self, 'x').size(0)
        else:
            return super().__inc__(key, value, *args, **kwargs)


class customTransformations(BaseTransform):
    def __call__(self, data: Data):

        new_data = customData.take_feat(data)
        return new_data


data_train = ZINC("/tmp/custom_zinc3", subset=True, split='train',
                  transform=Compose([customTransformations()]))

train_loader = DataLoader(data_train, batch_size=128)
batch = next(iter(train_loader))
print(batch)
```
works just fine for me.

**Reply by rusty1s** [reply]
But keep in mind that you are not really creating a `customData` instance here. You need to do:
```python
data = cls.from_dict(data.to_dict())
```
in the `take_feat` method.

**Reply by vkovac2** [reply]
Hello @rusty1s . Sorry for this late response. Is it possible that it works when you put the "customTransformation" in "tranform" and doesn't work when you put it in "pre_transform"?

```py

from typing import Any

import torch

from torch_geometric.data import Data
from torch_geometric.datasets import QM9, ZINC
from torch_geometric.loader import DataLoader
from torch_geometric.transforms import BaseTransform, Compose


class customData(Data):
    @classmethod
    def take_feat(cls, data: Data):

        relations = {
            0: [2, 7, 6, 5, 4, 3],
            1: [2, 5, 4, 3],
            2: [2, 3, 4, 5, 1]
        }
        data["relations"] = torch.nested.nested_tensor(list(
            relations.values()))
        data = cls.from_dict(data.to_dict())

        return data

    def __cat_dim__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'edge_index' in key:
            return 1
        else:
            return 0

    def __inc__(self, key: str, value: Any, *args, **kwargs) -> Any:
        if 'edge_index' in key:
            return torch.tensor([[getattr(self, 'x').size(0)],
                                 [getattr(self, 'x').size(0)]])
        if 'relation' in key:
            return getattr(self, 'x').size(0)
        else:
            return super().__inc__(key, value, *args, **kwargs)


class customTransformations(BaseTransform):
    def __call__(self, data: Data):

        new_data = customData.take_feat(data)
        return new_data

data_train = ZINC("/tmp/custom_zinc10000", subset=True, split='train',
                  pre_transform=Compose([customTransformations()]))

train_loader = DataLoader(data_train, batch_size=128)
batch = next(iter(train_loader))
print(batch)


```

Error I am getting is the following: 

```py
Exception has occurred: RuntimeError

Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
  File "/Users/try_zinc.py", line 101, in <module>
    data_train = ZINC("/tmp/custom_zinc10000", subset=True, split='train',

```


Thank you a lot for your time.

**Reply by vkovac2** [reply]
Hello @rusty1s. It seems that the PyTorch issue has been solved by [/pytorch/pytorch/pull/110219]. 

Thanks
